### PR TITLE
Suggestions for Annotation Processor compatibility

### PR DESCRIPTION
**Pull request by graemerocher**
This PR provides my suggestions and ideas to make the proposed CDI lite extension model annotation processor compatible.

Note that I have not run the tests and this PR merely serves to communicate my ideas and thoughts on the proposal in terms of Micronaut achieving compatibility with the API at some point in the future.

I have provided comments in the code with additional thoughts.

Changes are summarized below

## Alter API to be compatible with push based model of annotation processors

Prior to this change a pull based model of using *Query interfaces to retrieve *Info objects was in place. This will not work within the context of annotation processor which is event based and uses a push model. This change alters the API such that you register hooks through AnnotationConfigurable that can be called by the processor when an triggered.

For consistency DeclarationInfoProcessor is introduced as well for the synthesize phase.

## Remove Annotation based enhancements and use simple service loader model

Within the context of an annotation processor you don't want to slow down the compiler by having to scan all extensions in search of methods that are annotated with @Enhancement etc. It is better to use a simple ServiceLoader model and an interface to register extensions that doesn't add complexity to enhancement discovery.

## Rename Errors to Messages and Include in all Phases

In the annotation processor model you can and should report errors against the originating element so that the errors are propagated to javac. You can also report warnings and information so I have renamed Errors to Messages and allowed access it to the various phases such that errors can be reported in the appropriate phase and compilation can fail earlier if necessary.

## Expose `Types` interface for String-based API

The `Type` interface already includes a way to create a `Type` based on just a name of the type. It makes sense to expose this API and use it as a factory for the stringly typed API for both queries and also for defining annotation values that may reference a class not on the classpath.

### CODE REVIEW

File: independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/declarations/DeclarationInfo.java [hunk 1]
```diff
@@ -3,6 +3,17 @@
 import cdi.lite.extension.model.AnnotationTarget;
 import cdi.lite.extension.model.types.Type;
 
+// Graeme comments start
+// -------------
+// What I don't understand regarding the design on this interface
+// is why the need for the Kind and various isXXX() methods
+// plus methods that throw IllegalStateException.
+//
+// Surely it is better to use inheritance and instanceof checks plus casts.
```
Comment (Ladicek) [inline independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/declarations/DeclarationInfo.java:12]: I'm not exactly happy with this design either, but using `instanceof` and explicit casts certainly isn't "surely better".

File: independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/AppArchive.java [hunk 2]
```diff
@@ -4,11 +4,29 @@
...
-import java.util.Collection;
-import java.util.stream.Stream;
 
 public interface AppArchive {
+    // Graeme comments start
+    // -------------
+    // Enable access to the factory for creating types from strings.
+    // -------------
+    // Graeme comments end
+    Types types();
```
Comment (Ladicek) [inline independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/AppArchive.java:17]: If we go with the `ServiceLoader`, I'd rather put this into the `*Context` objects for various phases.

File: independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/Extension.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+     * Allows registering synthetic beans and observers.
+
+     * @param context The {@link SynthesisContext}
+     */
+    default void synthesize(SynthesisContext context) {}
+
+    // Graeme comments start
+    // -------------
+    // There is no end user hook for this phase, but I guess if you add a
+    // customizable synthesis phase this would be necessary.
```
Comment (Ladicek) [inline independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/Extension.java:76]: Yea the only need for separate phase here is to make sure it runs after all customization in "synthesis".

File: independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/AnnotationAttribute.java [hunk 4]
```diff
@@ -1,5 +1,20 @@
 package cdi.lite.extension.model;
 
+// Graeme comments start
+// -------------
+// In general I think the design of this API should be revisited, it seems to
+// be adapting Jandex which is quite ugly.
+//
+// In Micronaut we have AnnotationValue (roughly equivalent to AnnotationInfo) and
+// AnnotationValueBuilder which provide a fluent API and not this varargs approach
+// contained in this interface which seems much more verbose.
```
Comment (Ladicek) [inline independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/AnnotationAttribute.java:10]: I think you're talking about `AnnotationConfig` here, which certainly could be improved. Using a builder style would probably work better there, sure. It would probably let us get rid of the `Annotations` interface, which would be nice.

File: independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/AnnotationTarget.java [hunk 5]
```diff
@@ -5,6 +5,18 @@
 import java.lang.annotation.Annotation;
 import java.util.Collection;
 
+// Graeme comments start
+// -------------
+// This API seems a much more limited version of Micronaut's AnnotationMetadata and AnnotationSource APIs:
+//
+// https://docs.micronaut.io/latest/api/io/micronaut/core/annotation/AnnotationMetadata.html
+// https://docs.micronaut.io/latest/api/io/micronaut/core/annotation/AnnotationSource.html
```
Comment (Ladicek) [inline independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/AnnotationTarget.java:13]: `AnnotationTarget` from this proposal seems close to Micronaut's `AnnotationSource`. Also, `AnnotationInfo` from this proposal seems close to Micronaut's `AnnotationValue`. There's doesn't seem to be anything quite like `AnnotationMetadata`.

File: independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/types/Type.java [hunk 6]
```diff
@@ -3,6 +3,20 @@
...
 import cdi.lite.extension.model.declarations.DeclarationInfo;
 
+// Graeme comments start
+// -------------
+// Needs to be expanded to handle records.
+//
+// Also within Micronaut since we process source code what happens is when the AST is
+// visited we pass the elements that have already been resolved generics wise so there
+// is no need to check whether something is a wild card or a type variable or whatever.
+// I'm not sure the use case for needing this information nor the complexity it adds to this type.
```
Comment (Ladicek) [inline independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/types/Type.java:13]: I'm also not sure to what depth we need to model the Java type system (including things like being able to navigate from inner class to its containing class etc.), but I think we need to be able to inspect field types or method return/parameter types to certain degree. What exactly this "certain degree" is is surely up for debate.

File: independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/phases/synthesis/SyntheticObserverBuilder.java [hunk 7]
```diff
@@ -7,6 +7,15 @@
 import java.lang.annotation.Annotation;
 import javax.enterprise.event.TransactionPhase;
 
+// Graeme comments start
+// -------------
+// The behaviour of this needs more clarity:
+// * What does the observer type being added look like?
+// * What does typeWithTransitiveClosure do?
+// * Need javadoc / explanation / example
+// * Many TODOs here
```
Comment (Ladicek) [inline independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/phases/synthesis/SyntheticObserverBuilder.java:16]: This is definitely sketchy and needs a lot more work. The most important being, how to define the observer function (or construction/destruction functions in case of synthetic beans).

**Comment by Ladicek** [comment]
Hi @graemerocher, sorry for not replying sooner, I had to finish some other work plus this is quite big and I had to think about it for a while. Here are my thoughts on the major topics, plus I'll also add a few comments directly on the source where it's just minor details.

# Rename Errors to Messages and Include in all Phases

This is quite uncontroversial, except I don't really feel like we should design yet another logging facade. The original use case is custom validation of the bean deployment. In CDI Portable Extensions, validation errors are always represented by exceptions, so even my original proposal of `Errors` expands the scope a bit. I'm not sure how I feel about expanding it even more, though being able to add validation errors in all extension processing phases is probably a good idea.

# Remove Annotation based enhancements and use simple service loader model

We actually also considered basing the API on `ServiceLoader`. My personal opinion is that it's quite a lot less user-friendly. The biggest advantage of "extension is just an annotated method" is that the user (= extension author) can declare arbitrary parameters, and we provide them. Those parameters can then use annotations (or type parameters :-) ) to express simple constraints (e.g. `@WithAnnotations(MyAnnotation.class) ClassConfig<MyService>`, or some other syntax). Simple extensions are simple, complex extensions are complex.

With the approach you propose, you lose that flexibility and require users to use the complex API even for simple use cases. Other than that, obviously, there's no conceptual hurdle.

We could perhaps meet half-way. CDI Portable Extensions require implementing the `Extension` interface, and implementations are looked up via `ServiceLoader`. However, the interface doesn't prescribe any methods, they are looked up on the class based on annotations. We could do the same thing. Looking up the extension classes becomes cheap, and looking up a few annotated methods on a class shouldn't be expensive either.

# Removed the ability to look up types

I see you removed the annotations `@ExactType`, `@SubtypesOf` and `@SupertypesOf`. Those (or equivalent) are pretty important. In my original proposal, there are 2 basic ways to constrain which classes/methods/fields you want to configure: based on annotations, and based on which classes are they declared on. (The `*Query` interfaces then provide some additional constraints for more complex use cases.)

For example, in the SmallRye Fault Tolerance extension, we need to add a meta-annotation to all MicroProfile Fault Tolerance annotations (to establish the interceptor binding). That's a small set, so we can just directly declare a few parameters (using `@ExactType`, or using type arguments in my original proposal, or any other way) and directly modify them. No annotations involved.

I understand this might be a pain point for annotation processors, as they are triggered on annotation presence. Some annotation would often be present (e.g. `@Qualifier` in the Fault Tolerance case), but I'm not sure if it's always the case. This probably needs a lot more discussion.

# Alter API to be compatible with push based model of annotation processors

This change actually brings the API quite a lot closer to what Quarkus (ArC) does :-) I used the pull model, because I find it much more user-friendly (and had to build a translation layer to the Quarkus push model), though I understand that this is not a viable route. That said, I think a lot of user-friendliness can be retained for simple cases. Assuming we keep the "extensions are just annotated methods with arbitrary parameters" style, the idea is:

- No direct access to the various `*Info` classes is provided, they can only be accessed as described below. This means that extensions can't decare `*Info` parameters, just `*Config`, and `AppArchive*` no longer return `*Info` types, just `*Config`.
- `ClassConfig` no longer extends `ClassInfo` and no longer provides access to details about the class. (Same for fields and methods.) Instead, it only provides configuration methods.
    - `ClassConfig.with` (better name surely exists), which takes a callback and registers it to be executed later (when you get to process the class in question). Something like this:

        ```java
        @Enhancement
        // _somehow_ express that we should provide a ClassConfig for the MyFoo class,
        // different syntax is possible
        public void extension(ClassConfig<MyFoo> foo) {
            foo.with(config -> {
                // here, `config` provides access to `ClassInfo` of the `MyFoo` class, plus transformation methods
                // obviously the type of `config` must be different from the type of the `foo` parameter
                // (one can be still called `ClassConfig`, the other needs a different name)
            });
        }
        ```

    - `ClassConfig.addAnnotation(...)` and `ClassConfig.removeAnnotation(...)`, which are can be easily expressed using the `with` method and is just a convenience layer for simple use cases. These could be provided as `default` methods actually:

        ```java
        interface ClassConfig {
            void with(ClassConfigCallback callback);

            default void addAnnotation(Annotation annotation) {
                with(config -> config.addAnnotation(annotation));
            }

            // etc. etc. all transformation methods on `ClassConfigCallback` would be provided here in the same way
        }
        ```


- The `AppArchive` imperative entrypoint provides access to the same `*Config` types, and nothing more. Which means that they would also just register a callback to be executed later, perhaps with additional constraints that can't be expressed using parameters. For example:

    ```java
    @Enhancement
    public void extension(AppArchive app) {
        app.classes()
            .exactType(...)
            .subtypesOf(...)
            .annotatedWith(...)
            .annotatedWith(...)
            .configure(config -> {
                // again, `config` provides access to `ClassInfo` of the class, plus transformation methods
            });
    }
    ```

Basically, on the 1st level, you only get to register configuration callbacks, and on the 2nd level, in those callbacks, you can inspect the class/method/field as you wish. We would also specify that the `config` parameter passed into the callback (and everything accessible through it) is only valid during the execution of the callback, so storing it into a `static` field or something like that is illegal. (Implementations should be able to check for that relatively easily.)

This should retain the nature of my original proposal (and hence also most of its user-friendliness), while respecting all the constraints you have.

If we don't keep the "extensions are just annotated methods with arbitrary parameters" style, the idea should still mostly apply.

WDYT?

**Review by Ladicek — COMMENTED** [review]
Adding a few minor comments that don't deserve extra treatment :-)

**Comment by Ladicek** [comment]
One more thing. I'm not going to modify the code yet, as I feel like we'll want to do a few back and forths on this.

**Comment by graemerocher** [comment]
> Rename Errors to Messages and Include in all Phases

> This is quite uncontroversial, except I don't really feel like we should design yet another logging facade. The original use case is custom validation of the bean deployment. In CDI Portable Extensions, validation errors are always represented by exceptions, so even my original proposal of Errors expands the scope a bit. I'm not sure how I feel about expanding it even more, though being able to add validation errors in all extension processing phases is probably a good idea.
Remove Annotation based enhancements and use simple service loader model

It is not so much logging, let me give some background. For errors the user has to take action to continue since `javac` will fail compilation. Warnings are for deprecations you can pick up and warn the user about. Info messages is maybe the one you could debate about. What we use this for is for example projects like https://github.com/micronaut-projects/micronaut-openapi/ which generate a `swagger.yml` file at compilation time to inform the user where the file is. Whether that fits into this extension model is maybe up for debate. However if we only get errors I guess we could always have a framework specific way to output the other 2.



> We could perhaps meet half-way. CDI Portable Extensions require implementing the Extension interface, and implementations are looked up via ServiceLoader. However, the interface doesn't prescribe any methods, they are looked up on the class based on annotations. 

I'm fine with this, as you say what has to to be simple is the lookup of the classes, resolving the methods via reflection won't be that expensive.

> I see you removed the annotations @ExactType, @SubtypesOf and @SupertypesOf. Those (or equivalent) are pretty important. In my original proposal, there are 2 basic ways to constrain which classes/methods/fields you want to configure: based on annotations, and based on which classes are they declared on. (The *Query interfaces then provide some additional constraints for more complex use cases.)

> For example, in the SmallRye Fault Tolerance extension, we need to add a meta-annotation to all MicroProfile Fault Tolerance annotations (to establish the interceptor binding). That's a small set, so we can just directly declare a few parameters (using @ExactType, or using type arguments in my original proposal, or any other way) and directly modify them. No annotations involved.

I removed it because I wanted to simplify to just service loader, but ok I can see how this could be useful.


> Basically, on the 1st level, you only get to register configuration callbacks, and on the 2nd level, in those callbacks, you can inspect the class/method/field as you wish. We would also specify that the config parameter passed into the callback (and everything accessible through it) is only valid during the execution of the callback, so storing it into a static field or something like that is illegal. (Implementations should be able to check for that relatively easily.)

> This should retain the nature of my original proposal (and hence also most of its user-friendliness), while respecting all the constraints you have.

I like all these ideas and it is a step in the right direction. For a syntax like:

```java
public void extension(ClassConfig<MyFoo> foo) {
   ...
}
```

We would need to default to at least one annotation on that type for annotation processors, this could be `@Dependent` or some bean defining annotation.

Regarding the push model it is required from an annotation processing perspective since whatever gets generated needs to also report its originating element (the method,field,class that triggered the generation of new byte code) so that incremental compilation continues to work.

File: independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/types/Type.java (hunk 6 above)
Comment (graemerocher) [reply independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/types/Type.java:13]: Yes Micronaut doesn't model the whole system, just the basics of inspecting fields, arguments, return types etc.

File: independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/AnnotationTarget.java (hunk 5 above)
Comment (graemerocher) [reply independent-projects/arc/cdi-lite-ext-api/src/main/java/cdi/lite/extension/model/AnnotationTarget.java:13]: Yes the sum of all the `AnnotationMetadata` instances is equivalent to Quarkus' jandex index I would say

**Comment by Ladicek** [comment]
I think I'm personally fine with adding warn/info "logging". It could be seen as expanding the scope of CDI extensions a bit too much, but perhaps not, and I can see the utility. And it would be easy to back off, so we can add it to the proposal, sure.

> We would need to default to at least one annotation on that type for annotation processors, this could be @Dependent or some bean defining annotation.

That's what I feared. Current CDI Portable Extensions don't have that restriction, and neither does my original proposal. That said, I understand the requirement, and annotations often _are_ present. I just currently don't see a way how to force extension authors to specify which annotation to expect.

Continuing with my Fault Tolerance example, I declare a `ClassConfig<org.eclipse.microprofile.faulttolerance.Fallback>` so that I can "add" our (SmallRye-specific) interceptor binding annotation to the (MicroProfile standard) `@Fallback` annotation. It doesn't (and can't) have a bean defining annotation, but it sure does have the `@InterceptorBinding` annotation. So there _is_ something the annotation processor can catch. It can be specified as `@WithAnnotations(javax.interceptor.InterceptorBinding.class) ClassConfig<org.eclipse.microprofile.faulttolerance.Fallback>`, but I don't see a way how to force extension authors to always specify that.

Perhaps a different syntax is possible?

I don't know... perhaps get rid of the type parameter and put type constraints and annotation constraints into a single annotation?

```java
@Constraint(presentAnnotations = javax.interceptor.InterceptorBinding.class, exactType = org.eclipse.microprofile.faulttolerance.Fallback.class, subtypesOf = ..., supertypesOf = ...)
```

Actually annotation attributes can't have a `null` value, so a single annotation doesn't cut it. But we could retain `@ExactType`, `@SubtypesOf` and `@SupertypesOf`, but add a [mandatory] attribute `presentAnnotations`.

```java
@ExactType(type = org.eclipse.microprofile.faulttolerance.Fallback.class, presentAnnotations = javax.interceptor.InterceptorBinding.class)
```

Would something like that work?

__EDIT__: we'd have to do something similar for the `AppArchive` entrypoint. Perhaps instead of `app.classes().annotatedWith(...)`, where it's easy to avoid setting the expected annotations, do `app.annotatedClasses(...)`, where you can't escape it.

__EDIT__: I also didn't quite think about parameters of type `Collection<MethodConfig>` and similar; they don't make sense anymore. We could still have one `MethodConfig`, which would register a callback that would be called for all methods (satisfying given constraints). That's probably the best we can do. Or keep only classes, and allow transforming methods and fields in a class config callback.

> Regarding the push model it is required from an annotation processing perspective

Yea I understand that. It limits the expressive power quite a bit, unfortunately, but original CDI Portable Extensions are also push-based, so it shouldn't be a big deal conceptually.

Yesterday, I wrote that "We would also specify that the config parameter passed into the callback (and everything accessible through it) is only valid during the execution of the callback, so storing it into a static field or something like that is illegal". I've since realized that we'd probably have to relax this a bit.

In Portable Extensions, it is often required to build up some knowledge (in instance fields) during various callbacks, and later act on that knowledge. That isn't possible if we can't capture that knowledge in the callbacks.

(For example, I've recently added an annotation `@CircuitBreakerName` to SmallRye Fault Tolerance, which serves two purposes: provide a name for a `@CircuitBreaker` method, and serve as a qualifier for `@Inject`-ing a maintenance facade. Because names are just strings, I added validation. If no circuit breaker has a name `xyz` and yet someone asks to inject maintenance facade for `xyz`, deployment needs to fail during validation. I collect all existing names and also named injection points in an early Portable Extension phase, and during later phase, I check those two sets -- if they are distinct, it's a failure.)

This assumes that there's always exactly one instance of the extension. This is easy in Quarkus, but I assume that with annotation processing (and incremental compilation), you can invoke the extension multiple times, each time in a different JVM and on a different piece of the application. The assumption just doesn't hold, right?

Do you actually even have a point in time during which you have access to the entire application? I can't see doing validations like above without such access.

**Comment by graemerocher** [comment]
> I think I'm personally fine with adding warn/info "logging". It could be seen as expanding the scope of CDI extensions a bit too much, but perhaps not, and I can see the utility. And it would be easy to back off, so we can add it to the proposal, sure.

Great.

> That's what I feared. Current CDI Portable Extensions don't have that restriction, and neither does my original proposal. That said, I understand the requirement, and annotations often are present. I just currently don't see a way how to force extension authors to specify which annotation to expect.

> Continuing with my Fault Tolerance example, I declare a ClassConfig<org.eclipse.microprofile.faulttolerance.Fallback> so that I can "add" our (SmallRye-specific) interceptor binding annotation to the (MicroProfile standard) @Fallback annotation. 

Maybe I mispoke, the annotation doesn't have to be bean defining since you can use any annotation to "make it a bean". I see no issue with implementing the `@Fallback` example you provided from a Micronaut perspective. We would just process the `@Fallback` annotation and add the necessary beans based on it.

Also when I say there needs to be at least one annotation to trigger processing, this annotation can be anywhere. For example:

```java
class Foo {
     @Inject Bar bar;
}
```

Processing here is triggered by `@Inject` being present on field `bar`, there is no type level annotation indicating this a bean.

What we can't process is:

```java
class Foo {
    Bar bar;
}
```

Since there is no "trigger" annotation, although we could fall back to just doing `new Foo()` at an injection point where this type is injected which would pretty much give the same behaviour as CDI I believe.


> Yesterday, I wrote that "We would also specify that the config parameter passed into the callback (and everything accessible through it) is only valid during the execution of the callback, so storing it into a static field or something like that is illegal". I've since realized that we'd probably have to relax this a bit.

> In Portable Extensions, it is often required to build up some knowledge (in instance fields) during various callbacks, and later act on that knowledge. That isn't possible if we can't capture that knowledge in the callbacks.

yes we would need this knowledge too

> This assumes that there's always exactly one instance of the extension. This is easy in Quarkus, but I assume that with annotation processing (and incremental compilation), you can invoke the extension multiple times, each time in a different JVM and on a different piece of the application. The assumption just doesn't hold, right?

The annotation processor can hold singleton references to extensions, I don't see this is a problem, also for the validation phase that can be invoked when `processingEnvironment.processingOver()` returns true, which is at the end when everything has been built up.

> Do you actually even have a point in time during which you have access to the entire application? I can't see doing validations like above without such access.

There is a point in time where you can access all the generated beans and their injection points that were generated during annotation processing. This would allow you to implement the use case you mentioned above once `processingEnvironment.processingOver()` returned true. 

It may be technically possible to also load from the classpath somehow the beans on the classpath in other libraries, however I would prefer to avoid this as whilst it is technically possible to load the AST of already compiled classes in an annotation processor it would slow down the compiler significantly. I discussed this with @n1hility and @emmanuelbernard as well and they indicated they also discouraged loading the "whole world" as it is expensive for Quarkus as well.

**Comment by Ladicek** [comment]
> Maybe I mispoke, the annotation doesn't have to be bean defining since you can use any annotation to "make it a bean". I see no issue with implementing the @Fallback example you provided from a Micronaut perspective. We would just process the @Fallback annotation and add the necessary beans based on it.

It's a little strange example, but we actually need to "modify" (add an extra meta-annotation) the `@Fallback` annotation itself.

I guess that shouldn't be a problem for you -- I'm not sure how you implement annotation transformations currently, but I expect it's similar to Quarkus, where we have an extra layer (I call it "overlay") on top of the original immutable source of annotations (which in our case is Jandex and in your case is the AST). If yes, you'd just store the modification to your overlay and perhaps wouldn't even have to load the `Fallback` annotation class (depending on if we go with stringly-typed API or strongly-typed API).

> There is a point in time where you can access all the generated beans and their injection points that were generated during annotation processing.

Good!

I originally attempted to design the API so that it isn't necessary to build up knowledge in instance fields and act on it later -- instead, I wanted the extensions to be able to access all the required knowledge in one go. That's (another) reason for extensions being able to declare arbitrary parameters.

I'm not sure if that's still possible with the push-based model.

That said, later phases (I'm thinking synthesis and validation) probably need to be invoked exactly at that point where you have access to all the generated beans and injection points, so they probably don't have to be push-based. In my original proposal, these phases allow declaring parameters of type `ClassInfo` etc., which would probably have to be restricted. They could declare parameters of type `BeanInfo` and similar (and only be able to access types from there?), which should be enough I guess.

> It may be technically possible to also load from the classpath somehow the beans on the classpath in other libraries, however I would prefer to avoid this as whilst it is technically possible to load the AST of already compiled classes in an annotation processor it would slow down the compiler significantly.

Understood. I think we should be able to narrow down the API so that operations like this are either not possible, or restricted.

Perhaps it's time for me to get back to the API and try incorporating all the changes we debated here. I should also expand my prototype implementation to cover other phases (thinking discovery and validation at this point, synthesis is hard :-) ).

**Comment by graemerocher** [comment]
> I guess that shouldn't be a problem for you -- I'm not sure how you implement annotation transformations currently, but I expect it's similar to Quarkus, where we have an extra layer (I call it "overlay") on top of the original immutable source of annotations (which in our case is Jandex and in your case is the AST). If yes, you'd just store the modification to your overlay and perhaps wouldn't even have to load the Fallback annotation class (depending on if we go with stringly-typed API or strongly-typed API).

This is correct we have a kind of overlay, this is `AnnotationMetadata` and we don't need the annotations themselves at runtime. There are various APIs in Micronaut to perform annotation transformation like tasks such as:

* `AnnotationTransformer` (https://github.com/micronaut-projects/micronaut-core/blob/2.1.x/inject/src/main/java/io/micronaut/inject/annotation/AnnotationTransformer.java#L45) - Take an annotation and map to some other annotations and discard the original
* `AnnotationRemapper` (https://github.com/micronaut-projects/micronaut-core/blob/2.1.x/inject/src/main/java/io/micronaut/inject/annotation/AnnotationRemapper.java#L54) - Take an annotation in a package and remap it to another package (this is how we handle `jakarta.*` to `javax.*`
* `AnnotationMapper (https://github.com/micronaut-projects/micronaut-core/blob/2.1.x/inject/src/main/java/io/micronaut/inject/annotation/AnnotationMapper.java#L49) - Map one annotation to another but keep the other
* finally for more complex cases `TypeElementVisitor` (https://github.com/micronaut-projects/micronaut-core/blob/2.1.x/inject/src/main/java/io/micronaut/inject/visitor/TypeElementVisitor.java#L39) which has an `annotate(..)` method for imperative style which you can invoke in any class, method or field.

It is not too much of a stretch to map the concepts in this proposal to the above APIs.

> Perhaps it's time for me to get back to the API and try incorporating all the changes we debated here. I should also expand my prototype implementation to cover other phases (thinking discovery and validation at this point, synthesis is hard :-) ).

Sounds good 👍

**Comment by Ladicek** [comment]
Sorry for not getting back to this sooner. I just pushed https://github.com/Ladicek/quarkus-fork/commits/experiment-cdi-lite-ext-v2 which attempts to move from pull-based API to push-based API. I've mostly neglected all other concerns that we discussed, because I'm not exactly sure if the API is, in the end, as pleasant to use as I hoped. At this point, I'm thinking your original API might be a bit better, and perhaps even better would be to use the original Portable Extensions principle, where you don't register callbacks programmatically, but each "observer" method declared on the extension class is itself a callback. I'm interested in your thoughts.

**Comment by graemerocher** [comment]
Thanks hopefully I will get time to look next week although it is a busy one with several conference talks :)

**Comment by Ladicek** [comment]
Thanks, and no worries. I'll try to think about the API some more this week.

**Comment by Ladicek** [comment]
I've played a little with somewhat different approach for annotation transformation callbacks. It's here: https://github.com/Ladicek/quarkus-fork/commits/experiment-cdi-lite-ext-v3

The main difference compared to the first attempt is that annotation transformation callbacks are either declared or registered programmatically.

The `@Enhancement` methods are required to declare exactly one parameter of a type from the following set:

- `ClassConfig`, `MethodConfig` or `FieldConfig`: the method itself is an annotation transformation callback;
- `AppArchiveConfg`: the method is executed to register annotation transformation callbacks programmatically.

The `@ExactType` and `@SubtypesOf` annotations can now be placed on methods and are used in the declarative approach to constraint for which types (and with which annotations) is the method called.

The `@Enhancement` methods can still declare parameters of "utility" types, if that proves useful.

**Comment by graemerocher** [comment]
@Ladicek ok, will take a look this week, promise.

**Comment by Ladicek** [comment]
No worries, and thanks! :-)

**Comment by graemerocher** [comment]
Looks good to me and the current `SmallRyeFaultToleranceCdiLiteExtension` is totally implementable in Micronaut today just with different APIs, it would be relatively easy to support this API.

What is more vague for me is the `@Synthesis` phase and an example of how it would be used. Is there another extension that demonstrates how this would be used?

**Comment by graemerocher** [comment]
One note is that it should be possible for the smallrye extensions to ship with the metadata needed to make it a bean already computed, thus the extension could wouldn't need to run again in the end user application which just seems wasteful to me.

**Comment by Ladicek** [comment]
Thanks, that's great to know. So far, I've mostly focused on the `@Enhancement` stage (= annotation transformations). I could easily add a basic example of `@Discovery` to the SmallRye Fault Tolerance example (where we add beans that are not discovered automatically). I could also add an example of `@Validation` for Fault Tolerance, we do validate a few things. I didn't implement these phases yet, and the API proposal is really a sketch, but these 2 phases are rather straightforward.

I don't have an example of `@Synthesis`, I find synthetic beans/observers a more advanced use case. @mkouba @manovotn do you have an example or two of common situations that are solved by creating synthetic components?

**Comment by Ladicek** [comment]
One more thing -- I think it's possible, to certain degree, if I ignore extensions registering custom qualifiers etc., to precompute beans present in an archive. But I don't think you can possibly precompute what an extension does -- not without severely restricting its power.

My favorite example -- interceptor binding in SmallRye Fault Tolerance. It can either "modify" annotation classes from MicroProfile Fault Tolerance (by adding meta-annotations to them), or it can "modify" all methods and classes present in the application that use the MicroProfile Fault Tolerance annotations (by adding a SmallRye-specific annotation to them). In either case, the extension has to "change" classes outside the archive it's in.

**Comment by graemerocher** [comment]
Sure you can, you use case is completely implementable in Micronaut which does this today with no loss of power.

The way this works today in Micronaut would be something like:

```java
class SmallRyeFaultToleranceMapper implements TypedAnnotationMapper<Retry> {
    public List<AnnotationValue<?>> map(AnnotationValue<Retry> annotation, VisitorContext visitorContext) {
         return Collections.singletonList(
                AnnotationValue.builder("io.smallrye.faulttolerance.FaultToleranceBinding").build()
        );
    }
}
```

Which means whenever a `@Retry` annotation is found add `@FaulTolerenceBinding`. Now currently that would add it to the method, but if you wanted to do this on the type you would do:

```java
// visits only methods annotated with @Retry
class SmallRyeFaultToleranceVisitor implements TypeElementVisitor<Object, Retry> {
    ClassElement clazz;
    public void visitClass(ClassElement element, VisitorContext context) {
        this.clazz = classElement;
    }
   public void visitMethod(MethodElement element, VisitorContext context) {
          clazz.annotate("io.smallrye.faulttolerance.FaultToleranceBinding")
    }
}
```

Now for classes external to the application, we don't currently support visiting them dynanically (because the metadata is already precomputed in the jar file) but we could add that support and in fact users have been asking for this since there are some cases where the JAR file is not in the users control and they want to the produce DI metadata.

It is this last case where I think maybe `beans.xml` would be valuable as we would be to identify with that the jars to process in this way.

**Comment by Ladicek** [comment]
Ah OK -- I misunderstood what you meant by precomputation. If running extensions is part of the precomputation process, then that's totally fine.

**Comment by mkouba** [comment]
>  I find synthetic beans/observers a more advanced use case. @mkouba @manovotn do you have an example or two of common situations that are solved by creating synthetic components?

@Ladicek A typical example could be: I need to register a bean for every required type from an injection point annotated with a specific annotation where the set of required types is not known before (e.g. custom converters in MP Config?).

**Comment by manovotn** [comment]
> I don't have an example of @Synthesis, I find synthetic beans/observers a more advanced use case. @mkouba @manovotn do you have an example or two of common situations that are solved by creating synthetic components?

Providing a bean for a 3rd party JAR that doesn't use CDI. Assuming you cannot just register the bean class (because of, say, non trivial constructor), you will need to create a synthetic bean.

What Martin said is pretty common indeed. In more general terms, its basically "dynamic" (or rather "conditional") bean creation based on injection points within deployment.

**Comment by Ladicek** [comment]
FYI I've just pushed a prototype implementation, together with my favorite example of SmallRye Fault Tolerance extension, of the `AppArchiveBuilder` part of `@Discovery` to https://github.com/Ladicek/quarkus-fork/commits/experiment-cdi-lite-ext-v4

**Comment by Ladicek** [comment]
FYI, I've finished a prototype implementation of `@Discovery` (`AppArchiveBuilder` was there, now `Contexts` is as well, even though I have no test for it, it's a niche thing) and added prototype implementation of `@Validation` (where I've changed `AppDeployment` to also follow the push style).

I forgot that I only do custom validation for Fault Tolerance in the upcoming version, so I've added an example usage of `Messages` to https://github.com/Ladicek/quarkus-fork/blob/experiment-cdi-lite-ext-v4/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/cdi/lite/ext/ChangeQualifierTest.java

**Comment by Ladicek** [comment]
Also, I'm going to prototype `@Synthesis` next, so that we can get to something that works for all of us.

**Comment by Ladicek** [comment]
FYI I've just pushed a prototype implementation of `@Synthesis` into https://github.com/Ladicek/quarkus-fork/commits/experiment-cdi-lite-ext-v4, so now the complete proposed API is there (and can be experimented with). I also added 2 basic tests for synthetic beans and observers.
